### ISSUE TITLE
[Feature] Shell Conditions for all plugins

### ISSUE THREAD

**Issue by robobenklein** [anishathalye/dotbot#143]
I've seen a lot of people wanting to do things like optional tasks, or only run sets of tasks, machine specific configs, etc.

#96, #44 etc.

I think I might have some level of solution.

Here's how an implementation might look:

```yaml
# "Put this only on gaming machines"
- link:
    ~/.steam/someconfig:
      path: gaming/steam/config
      create: false
      condition: [[ -d ~/.steam ]]
```

Link plugin (and perhaps shell plugin) gets additional directive: `condition`,`shellcond`,`shellcondition` or whatever, which is a command that is executed and the return status is used as a boolean. If shell returns true (retval 0) then perform the operation, else don't.

#96 i3 Example:

```yaml
- shell:
  - sudo apt install i3
    condition: ! dpkg -s i3 # using ! as in https://stackoverflow.com/questions/367069/how-can-i-negate-the-return-value-of-a-process
- link:
  ~/.config/i3:
    path: i3
    condition: dpkg -s i3 # success if i3 is installed
```

#44 Shell vars Example:
```yaml
- link:
  ~/.somewhere:
    path: something
    condition: [[ ! -z "$INSTALL_SOMETHING" ]]
```
Could be run as `INSTALL_SOMETHING=1 ./install`

What's the feel for a solution somewhat like this?

**Comment by arcticicestudio** [comment]
Looks like this is related to #88.

**Comment by robobenklein** [comment]
Alright, I've mulled over this idea for long enough to try and implement it.

I need some opinions:

1. What should the directive be called? (`condition`, `shellcond`, `cond`, etc, other ideas?)
2. Should we use the user's shell to eval the commands? (I would think so...)
3. Does this belong in the shell or clean plugins as well? Or should I create something external to the plugins and have a generic function for boolean success?
4. Do we need anything more complex than just checking the return status of a command?

**Comment by robobenklein** [comment]
https://github.com/robobenklein/dotbot/commit/791a258f3e6b32b960cb13d9f9ca51db60482608

It's been working great for me, anyone else want to try and find bugs before I cherry a new PR?

**Comment by anishathalye** [comment]
Haven't had time to test it yet, but minor note from skimming through the code -- we might want to redirect stdout/stderr to devnull.

**Comment by robobenklein** [comment]
I redirect stderr to stdout here, and stdout gets captured by subprocess:

https://github.com/robobenklein/dotbot/commit/791a258f3e6b32b960cb13d9f9ca51db60482608#diff-99c86930d880d5300a214b25ff49337cR254

**Comment by anishathalye** [comment]
Do we want to show it at all, though?

**Comment by robobenklein** [comment]
I didn't think so...

If we wanted we could log.debug the osstdout? Although I'm not sure about multiline outputs and I think that has potential to get a bit spammy. Right now I thought that it was enough just to log the command itself when it failed.

Maybe this is another flag? Like `showoutput`? (or a better name)

What are the users' opinions on this? Have there been discussions on verbosity level?

**Comment by tony** [comment]
I'm a [dotfiles](https://github.com/jbernard/dotfiles) user that was about to fork/create a new project just to get this kind of behavior.

I'm trying to get conditional linking based on:

- Arbitrary machine-names
- Hostname
- Platform

I suppose shell command conditionals could do that.

(A second, unrelated thing was populating template variables in configs)

I've checked the PR's as of April 14th, 2018, and the opportunity to create a PR for this issue, or anything is still open at the moment.

> What should the directive be called? (condition, shellcond, cond, etc, other ideas?)

It could be ``if``. It's short, it saves space on the line, unambiguous, and it would make it somewhat intuitive the command only runs if x is true.

> Do we need anything more complex than just checking the return status of a command?

To make an initial simple, it can just be that.  Return status can be pretty flexible, since the conditions it wraps can be totally arbitrary.

ATM I use similar behavior in another project, which actually allows invoking an arbitrary script file and check the return status of it. And it's been well-received and pretty flexible.

Are you interested in making a PR ?

**Comment by anishathalye** [comment]
I'm fine with the general idea of this feature, so if @robobenklein, it'll get merged eventually.

**Comment by robobenklein** [comment]
Yeah, I can get a PR going. Originally I had stuck the shell conditions on top of the globbing, so I was just waiting to get globbing in before I made the one for conditionals.

I'll probably do some quick code cleanup and docs, and then it shouldn't be too long before I submit a PR.

**Comment by anishathalye** [comment]
A version of this is implemented in #169 / merged in 1b17e8e.
